### PR TITLE
Introduce QFileSystemListView & QFileSystemTreeView

### PR DESCRIPTION
**Pull request by pasnox**
Conveniant views with QFileSystemModel support.

### CODE REVIEW

File: qtwidgetsextra/filesystem/QFileSystemTreeViewPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    : QObject(parent)
+    , m_initialized(false)
+{
+}
+
+void QFileSystemTreeViewPlugin::initialize(QDesignerFormEditorInterface * /* core */)
+{
+    if (m_initialized) {
+        return;
+    }
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeViewPlugin.cpp]: Too empty lines

File: qtwidgetsextra/filesystem/QFileSystemTreeView.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    Q_PROPERTY(bool readOnly READ readOnly WRITE setReadOnly)
+    Q_PROPERTY(bool resolveSymlinks READ resolveSymlinks WRITE setResolveSymlinks)
+    Q_PROPERTY(QDir::Filters filters READ filters WRITE setFilters)
+    Q_PROPERTY(QStringList nameFilters READ nameFilters WRITE setNameFilters)
+    Q_PROPERTY(QFileSystemTreeView::Sections visibleSections READ visibleSections WRITE setVisibleSections)
+    Q_PROPERTY(QString rootPath READ rootPath WRITE setRootPath)
+
+public:
+    enum SectionFlag {
+        AllSections = 0,
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.h]: why using hex values?

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                    top = -1;
+                    bottom = -1;
+                }
+            }
+        }
+
+        widget->selectionModel()->select(itemSelection, flags);
+    }
+
+public:
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: Public is repeated for nothing

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            int bottom = -1;
+
+            for (int i = 0; i < indexes.count(); i++) {
+                const QModelIndex &index = indexes[i];
+                const bool hasNext = i +1 < indexes.count();
+
+                if (top ==-1) {
+                    top = index.row();
+                }
+                else {
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: five indentation level from the foreach to the last if/else. Can't do something better less deep?

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+void QFileSystemTreeView::setNameFilters(const QStringList &nameFilters)
+{
+    d->model->setNameFilters(nameFilters);
+}
+
+QFileSystemTreeView::Sections QFileSystemTreeView::visibleSections() const
+{
+    QFileSystemTreeView::Sections sections = QFileSystemTreeView::AllSections;
+
+    if (header()->hiddenSectionCount() > 0) {
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: Hum magic number!!

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        }
+    }
+
+    return sections;
+}
+
+void QFileSystemTreeView::setVisibleSections(QFileSystemTreeView::Sections sections)
+{
+    if (sections == QFileSystemTreeView::AllSections) {
+        header()->setSectionHidden(0, false);
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: magic number!!

File: qtwidgetsextra/filesystem/QFileSystemListViewPlugin.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    , m_initialized(false)
+{
+}
+
+void QFileSystemListViewPlugin::initialize(QDesignerFormEditorInterface * /* core */)
+{
+    if (m_initialized) {
+        return;
+    }
+
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListViewPlugin.cpp]: too empty lines

File: qtwidgetsextra/filesystem/QFileSystemListView.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    Q_PROPERTY(bool browsable READ browsable WRITE setBrowsable)
+    Q_PROPERTY(QDir::Filters filters READ filters WRITE setFilters)
+    Q_PROPERTY(QStringList nameFilters READ nameFilters WRITE setNameFilters)
+    Q_PROPERTY(QFileSystemListView::SectionFlag visibleSection READ visibleSection WRITE setVisibleSection)
+    Q_PROPERTY(QString rootPath READ rootPath WRITE setRootPath)
+
+public:
+    enum SectionFlag {
+        AllSections = 0,
+        NameSection = 0x1,
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListView.h]: Why using hex values?

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+        foreach (const QString &item, selection) {
+            const QModelIndex index = model->index(item, column);
+
+            if (index.isValid()) {
+                mapIndexes[index.parent()][index]++;
+            }
+        }
+
+        foreach (const QModelIndex &parent, mapIndexes.keys()) {
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: five indentation level from the foreach to the last if/else. Can't do better?

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    void activated(const QModelIndex &index) {
+        if (!widget->browsable() || !model->isDir(index) || QApplication::keyboardModifiers() != Qt::NoModifier) {
+            return;
+        }
+
+        const QString filePath = model->filePath(index);
+        widget->setRootPath(filePath);
+    }
+
+public:
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Public is repeated and some part where separated  by the public slots

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+void QFileSystemListView::setNameFilters(const QStringList &nameFilters)
+{
+    d->model->setNameFilters(nameFilters);
+}
+
+QFileSystemListView::SectionFlag QFileSystemListView::visibleSection() const
+{
+    switch (modelColumn()) {
+        case 0:
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Magic Number!!!

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        default:
+            return QFileSystemListView::NameSection;
+    }
+}
+
+void QFileSystemListView::setVisibleSection(QFileSystemListView::SectionFlag section)
+{
+    switch (section) {
+        case QFileSystemListView::NameSection:
+            setModelColumn(0);
```
Comment (Schoumi) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Magic Number

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 9 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Looks not :)

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 10 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Yeah I don't like much putting in the same public scope members and functions.
I will check Qt convention.

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 11 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Ok, I will add Column enum.

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 12 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemListView.cpp]: Same

File: qtwidgetsextra/filesystem/QFileSystemListView.h (hunk 8 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemListView.h]: That's or'derable flags.

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp (hunk 4 above)
Comment (pasnox) [inline qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: Not without loosing readibility I guess.

File: qtwidgetsextra/filesystem/QFileSystemTreeViewPlugin.cpp (hunk 1 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemTreeViewPlugin.cpp]: done

File: qtwidgetsextra/filesystem/QFileSystemTreeView.h (hunk 2 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemTreeView.h]: that' flags and easier to read

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp (hunk 5 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: done

File: qtwidgetsextra/filesystem/QFileSystemTreeView.cpp (hunk 6 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemTreeView.cpp]: done

File: qtwidgetsextra/filesystem/QFileSystemListViewPlugin.cpp (hunk 7 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemListViewPlugin.cpp]: done

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 11 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemListView.cpp]: done

File: qtwidgetsextra/filesystem/QFileSystemListView.cpp (hunk 12 above)
Comment (pasnox) [reply qtwidgetsextra/filesystem/QFileSystemListView.cpp]: done

**Pull request closed without merging**